designs/0030-complex-numbers.md

```diff
@@ -0,0 +1,355 @@
+
+The basic arithmetic and matrix operators extend to complex numbers as expected, given the reduction $i^2 = -1.$.  Other functions, such as sine and logarithm, have more complicated extensions to the complex domain.  C++ provides support for complex numbers through its `<complex>` header, which defines all of the basic mathematical operations on complex numbers.
+
+## Goals
+
+The goal of this feature is to allow users to code models involving complex numbers.  The reason we want to extend Stan to complex numbers is that many statistical problems are most naturally formulated in the complex domain.  For example, in imaging applications such as magnetic resonance imaging (MRI) and cryo electron microscopy (cryo-em), we do not observe an image, but only the magnitude (or modulus) of its discrete Fourier transform.  This is naturally a problem that requires a complex FFT operation.  Similarly, eigendecomposition of asymmetric matrices is fundamentally a complex operation, as is Schur decomposition.  Stationarity conditions for time-series models are often stated in terms of conditions on the magnitudes of complex roots.
+
+The goal of this document is to be precise enough that
+    - users can understand the feature well enough to evaluate it, and
+    - developers can work on a technical specification for the implementation.
+
+## Design considerations
+
+- The main design consideration is that Stan programs involving complex numbers be easy to write, and just as importantly, easy to read.
+
+- We should follow the basic design of Stan types, including promotion structure and associated covariance (in the CS not stats sense).
+
+- I/O structures need to
+    - balance self documentation and efficiency, and
+    - be easy to adapt to the interfaces and analysis tools.
+
+
+## Assumptions
+
+- The language will target C++ code in the Stan math library.
+
+
+## Stan data types
+
+The following four basic data types will handle all of the complex number functionality we need:
+
+- `complex`:  scalar type
+- `complex_vector[N]`: vector of complex values
+- `complex_row_vector[M]`: row vector of complex values
+- `complex_matrix[M, N]`: matrix of complex values
+
+As usual, arrays can be constructed from any of these types.
+
+### Constrained data types [Optional]
+
+Lower and upper bounds could be construed as bounds on absolute value (or modulus) of a complex number, with
+```
+complex<lower = 1> z;
+```
+requiring `modulus(z) > 1`, where $\mod x + yi = \sqrt{x^2 + y^2}.$  Upper bounds and mixed upper and lower bounds work the same way.  The bounds are real valued.
+
+Lower and upper bound constraints can be applied to any of the complex types.
+
+This feature is included for completeness, but marked "optional" because it is not clear it will be useful as realistic complex constraints are typically multivariate.
```

Comment:
This was not implemented

Reply:
I think it's OK to leave it as an optional feature.  I'm not sure it's useful, so also feel free to cut.

Reply:
Removed.